nvflare/app_common/tie/controller.py

```diff
@@ -0,0 +1,565 @@
+                return
+            if not isinstance(peer_ctx, FLContext):
+                self.log_error(fl_ctx, f"expect peer_ctx to be FLContext but got {type(peer_ctx)}")
+                return
+            client_name = peer_ctx.get_identity_name()
+            if not client_name:
+                self.log_error(fl_ctx, "missing identity from peer_ctx")
+                return
+            status = self.client_statuses.get(client_name)
+            if not status:
+                self.log_error(fl_ctx, f"no status record for client {client_name}")
+            assert isinstance(status, _ClientStatus)
+            if op:
+                status.last_op = op
+            if client_done:
+                status.app_done = client_done
+            status.last_op_time = time.time()
+
+    def _process_client_done(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        """Process the ClientDone report for a client
+
+        Args:
+            topic: topic of the message
+            request: request to be processed
+            fl_ctx: the FL context
+
+        Returns: reply to the client
+
+        """
+        self.log_debug(fl_ctx, f"_process_client_done {topic}")
+        exit_code = request.get(Constant.MSG_KEY_EXIT_CODE)
+
+        if exit_code == 0:
+            self.log_info(fl_ctx, f"app client is done with exit code {exit_code}")
+        elif exit_code == Constant.EXIT_CODE_CANT_START:
+            self.log_error(fl_ctx, f"app client failed to start (exit code {exit_code})")
+            self.system_panic("app client failed to start", fl_ctx)
+        else:
+            # Should we stop here?
+            # Problem is that even if the exit_code is not 0, we can't say the job failed.
+            self.log_warning(fl_ctx, f"app client is done with exit code {exit_code}")
+
+        self._update_client_status(fl_ctx, client_done=True)
+        return make_reply(ReturnCode.OK)
+
+    def _handle_app_request(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        """Handle app request from applets on other sites
+        It calls the connector to process the app request. If the connector fails to process the request, the
+        job will be stopped.
+
+        Args:
+            topic: message topic
+            request: the request data
+            fl_ctx: FL context
+
+        Returns: processing result as a Shareable object
+
+        """
+        self.log_debug(fl_ctx, f"_handle_app_request {topic}")
+        op = request.get_header(Constant.MSG_KEY_OP)
+        if self._is_stopped():
+            self.log_warning(fl_ctx, f"dropped app request ({op=}) since server is already stopped")
+            return make_reply(ReturnCode.SERVICE_UNAVAILABLE)
+
+        # since XGB protocol is very strict, we'll stop the control flow when any error occurs
+        process_error = "XGB request process error"
+        self._update_client_status(fl_ctx, op=op)
```

Comment:
A general thought, should we be monitoring and handling the client status in each controller?

Could some of these status be general that it is worth doing at a lower level or wrap as an common method/API of the base class Controller?

Reply:
Let's think about that. I can see that some of the client lifecycle management logic (e.g. no activity for an extended period of time) may apply to most/all controllers. In fact, similar logic already exists for xgb, swarm learning (or CWF), and TIE.

We should think about how to consolidate them into a base controller.